### PR TITLE
Feat: AMQP module

### PR DESCRIPTION
**Pull request by bradleytrf**
https://truefoundry.atlassian.net/browse/TT-6142

### CODE REVIEW

**Comment by bradleytrf** [comment]
313009717-407f9f29-6971-4c05-8b08-e75f50f587cb
313009891-31ca9060-7507-4458-9b0e-1a5b058c2184

File: async_processor/amqp_pub_sub.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        if self._queue:
+            return self._queue
+        await self._get_channel()
+        self._queue = await self._channel.declare_queue(self._queue_name, durable=True)
+        return self._queue
+
+    @asynccontextmanager
+    async def get_input_message(
+        self,
+    ) -> AsyncIterator[Optional[str]]:
```
Comment (akashg3627) [inline async_processor/amqp_pub_sub.py:48]: You can directly return bytes here. Check nats code once

File: async_processor/amqp_pub_sub.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        queue = await self._get_queue()
+        try:
+            message = await queue.get(fail=False, timeout=self._wait_time_seconds)
+        except Exception as ex:
+            raise InputMessageFetchFailure(f"Error fetch input message: {ex}") from ex
+        if not message:
+            yield None
+            return
+        try:
+            yield message.body.decode()
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:59]: You can directly yield bytes.

https://github.com/truefoundry/async_service/blob/c5d92379a631ebba71353e5a275304dc4285175a/async_processor/nats_pub_sub.py#L107

File: async_processor/amqp_pub_sub.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        self._channel = None
+        self._queue = None
+
+    async def _get_connect(self):
+        if self._connection:
+            return self._connection
+        self._connection = await connect_robust(self._queue_url)
+        return self._connection
+
+    async def _get_channel(self):
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:95]: Can you please add return type.

File: async_processor/amqp_pub_sub.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+class AMQPInput(Input):
+    def __init__(self, config: AMQPInputConfig):
+        self._queue_url = config.queue_url
+        self._queue_name = config.queue_name
+        self._wait_time_seconds = config.wait_time_seconds
+        self._connection = None
+        self._channel = None
+        self._queue = None
+
+    async def _get_connect(self):
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:25]: Can we please add return types to all these methods if they are returning something?

File: async_processor/amqp_pub_sub.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        return self._queue
+
+    @asynccontextmanager
+    async def get_input_message(
+        self,
+    ) -> AsyncIterator[Optional[bytes]]:
+        message = None
+        queue = await self._get_queue()
+        try:
+            message = await queue.get(fail=False, timeout=self._wait_time_seconds)
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:52]: What will happen at timeout?
Will there be an exception raised? 
If yes, do we need to handle it?

File: async_processor/amqp_pub_sub.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        if self._connection:
+            return self._connection
+        self._connection = await connect_robust(self._queue_url)
+        return self._connection
+
+    async def _get_channel(self) -> Channel:
+        if self._channel:
+            return self._channel
+        await self._get_connect()
+        self._channel = await self._connection.channel()
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:99]:
```suggestion
        connection = await self._get_connect()
        self._channel = await connection.channel()
```

File: async_processor/amqp_pub_sub.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        return self._channel
+
+    async def _get_queue(self):
+        if self._queue:
+            return self._queue
+        await self._get_channel()
+        self._queue = await self._channel.declare_queue(self._queue_name, durable=True)
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:106]:
```suggestion
        channel = await self._get_channel()
        self._queue = await channel.declare_queue(self._queue_name, durable=True)
```

File: async_processor/amqp_pub_sub.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        return self._queue
+
+    async def publish_output_message(
+        self, serialized_output_message: bytes, request_id: Optional[str]
+    ):
+        channel = await self._get_channel()
+        await self._get_queue()
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:113]: This seems odd to me. Why do we need to call it here?

File: async_processor/amqp_pub_sub.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        if self._connection:
+            return self._connection
+        self._connection = await connect_robust(self._queue_url)
+        return self._connection
+
+    async def _get_channel(self) -> Channel:
+        if self._channel:
+            return self._channel
+        await self._get_connect()
+        self._channel = await self._connection.channel()
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:35]:
```suggestion
        connection = await self._get_connect()
        self._channel = await connection.channel()
```

File: async_processor/amqp_pub_sub.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        return self._channel
+
+    async def _get_queue(self):
+        if self._queue:
+            return self._queue
+        await self._get_channel()
+        self._queue = await self._channel.declare_queue(self._queue_name, durable=True)
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:42]:
```suggestion
        channel = await self._get_channel()
        self._queue = await channel.declare_queue(self._queue_name, durable=True)
```

File: pyproject.toml [hunk 11]
```diff
@@ -26,8 +26,10 @@ cryptography = ">=41.0.6,<42.1.0"
 nats-py = { version = "2.6.0", optional = true }
 boto3 =  { version = ">=1.28.16,<1.29.0", optional = true }
 kafka-python = { version = ">=2.0.0,<2.1.0", optional = true }
+aio-pika = {version = "^9.4.0", optional = true}
```
Comment (debajyoti-truefoundry) [inline pyproject.toml:29]:
```suggestion
aio-pika = {version = ">=9.4.0,<10.0.0", optional = true}
```

File: async_processor/amqp_pub_sub.py (hunk 8 above)
Comment (bradleytrf) [reply async_processor/amqp_pub_sub.py:113]: Yes, I added this to create the queue
if we don't add this, the data will be skipped

File: pyproject.toml (hunk 11 above)
Comment (bradleytrf) [reply pyproject.toml:29]: is this manually added ?

**Comment by bradleytrf** [comment]
Screenshot 2024-03-15 at 08 23 12
tested on google rabbitmq

**Review by akashg3627 — APPROVED** [review]

File: async_processor/amqp_pub_sub.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+            raise InputMessageFetchFailure(f"Error fetch input message: {ex}") from ex
+        if not message:
+            yield None
+            return
+        try:
+            yield message.body
+        except Exception as ex:
+            raise InputMessageFetchFailure(
+                f"Error decoding input message body: {ex}"
+            ) from ex
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:68]: This exception handling is incorrect.

Comment (debajyoti-truefoundry) [reply async_processor/amqp_pub_sub.py:68]:
```suggestion
```

File: async_processor/amqp_pub_sub.py (hunk 8 above)
Comment (debajyoti-truefoundry) [reply async_processor/amqp_pub_sub.py:118]: Ideally, the queue should be created by the Ops/Developer.

Comment (debajyoti-truefoundry) [reply async_processor/amqp_pub_sub.py:118]: When we start, we should validate that the queue exists. Error out otherwise.
https://github.com/truefoundry/async_service/blob/c5d92379a631ebba71353e5a275304dc4285175a/async_processor/nats_pub_sub.py#L50

Comment (debajyoti-truefoundry) [reply async_processor/amqp_pub_sub.py:118]: I have added an example above for NATS.

File: async_processor/types.py [hunk 13]
```diff
@@ -216,6 +216,20 @@ def to_input(self) -> Input:
         return SQSInput(self)
 
 
+class AMQPInputConfig(InputConfig):
+    type: constr(regex=r"^amqp$") = "amqp"
+    queue_url: constr(
+        regex=r"^(?:amqp|amqps):\/\/(?:([^:/?#\s]+)(?::([^@/?#\s]+))?@)?([^/?#\s]+)(?::(\d+))?\/?([^?#\s]*)?(?:\?(.*))?$"
+    )
+    queue_name: str
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:224]: Do we need both queue URL and queue name?

File: async_processor/types.py [hunk 14]
```diff
@@ -216,6 +216,20 @@ def to_input(self) -> Input:
         return SQSInput(self)
 
 
+class AMQPInputConfig(InputConfig):
+    type: constr(regex=r"^amqp$") = "amqp"
+    queue_url: constr(
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:221]: Should this be called `queue_url`? If this is just the URL for rabbitmq, maybe we call it uri/url?

Comment (debajyoti-truefoundry) [reply async_processor/types.py:221]: Let us not call it queue URL as this URL is not queue specific.

**Comment by bradleytrf** [comment]
Screenshot 2024-03-18 at 09 32 32

File: async_processor/amqp_pub_sub.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+        self._queue = None
+
+    async def _validate_consumer_exists(self):
+        channel = await self._get_channel()
+        try:
+            self._queue = await channel.declare_queue(self._queue_name, passive=True)
+        except ChannelNotFoundEntity as ex:
+            raise Exception(
+                f"Consumer {self._queue_name!r} does not exist."
+                " Please create the consumer before running the async processor."
```
Comment (akashg3627) [inline async_processor/amqp_pub_sub.py:35]:
```suggestion
                " Please create the queue before running the async processor."
```

**Review by akashg3627 — APPROVED** [review]

File: async_processor/amqp_pub_sub.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+        except Exception:
+            logger.exception("Failed to drain and close nats connection")
+
+    async def publish_output_message(
+        self, serialized_output_message: bytes, request_id: Optional[str]
+    ):
+        channel = await self._get_channel()
+        await self._get_queue()
+        await channel.default_exchange.publish(
+            Message(body=serialized_output_message), routing_key=self._queue_name
```
Comment (chiragjn) [inline async_processor/amqp_pub_sub.py:160]: I see we have restricted ourselves to default exchange and re-using queue name as routing key

Shouldn't the exchange name and routing key be distinct inputs?
afaik, for publishing we need queue name, just the exchange + routing key
What queue is bound to what exchange and routing key is irrelevant when publishing

See: https://aio-pika.readthedocs.io/en/latest/rabbitmq-tutorial/4-routing.html#bindings

File: async_processor/amqp_pub_sub.py [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+                    f"Error publishing input message: {ex}"
+                ) from ex
+
+    async def publish_input_message(
+        self, serialized_input_message: bytes, request_id: str
+    ):
+        channel = await self._get_channel()
+        await channel.default_exchange.publish(
+            Message(body=serialized_input_message), routing_key=self._queue_name
+        )
```
Comment (chiragjn) [inline async_processor/amqp_pub_sub.py:101]: exchange name and queue name should be configurable
Maybe not today but some day in future?

File: async_processor/types.py [hunk 18]
```diff
@@ -288,6 +302,21 @@ def to_output(self) -> Output:
         return SQSOutput(self)
 
 
+class AMQPOutputConfig(OutputConfig):
+    type: constr(regex=r"^amqp$") = "amqp"
+
+    url: constr(
+        regex=r"^(?:amqp|amqps):\/\/(?:([^:/?#\s]+)(?::([^@/?#\s]+))?@)?([^/?#\s]+)(?::(\d+))?\/?([^?#\s]*)?(?:\?(.*))?$"
+    )
+    queue_name: str
```
Comment (chiragjn) [inline async_processor/types.py:311]: Ideally this should be

```suggestion
    exchange_name: str
    routing_key: str
```

File: async_processor/amqp_pub_sub.py (hunk 17 above)
Comment (chiragjn) [reply async_processor/amqp_pub_sub.py:101]: Another question - Is it always guaranteed that a routing key = X always delivers to a queue named X?

**Review by chiragjn — APPROVED** [review]
I am approving the PR because it is in functional state
But we should sync offline and see if we want to refactor output config to take exchange and routing key instead of queue name

File: async_processor/amqp_pub_sub.py [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+        try:
+            await self._nc.close()
+        except Exception:
+            logger.exception("Failed to drain and close nats connection")
+
+    async def publish_output_message(
+        self, serialized_output_message: bytes, request_id: Optional[str]
+    ):
+        channel = await self._get_channel()
+        await self._get_queue()
```
Comment (akashg3627) [inline async_processor/amqp_pub_sub.py:158]: why is this needed?
